### DISCUSSION TITLE
Type for not allowing NUL characters in strings

### DISCUSSION THREAD

**Discussion by Spaxterr** [pydantic/pydantic#9316] [Question] [answered]
Hello, I would like a type for my models that validates strings and makes sure they are not empty and do not contain any NUL (`\0`) characters. The closest I've gotten is `Field(min_length=1, strip_whitespace=True, pattern='[^\0]+')`. However I'd like to bundle this into a type rather than using `Field` on every attribute.

I'd like something like this if possible:
```py
class MyModel(BaseModel)
    value: NonEmptyString  # Checks for empty and NUL characters
```

**Comment by ybressler** [comment, marked as answer]
You can use a constrained string:
```py3
from pydantic import constr
NonEmptyString = constr(strip_whitespace=True, min_length=1, pattern='[^\0]+')
```
Alternatively, if you want more control, you can use an [annotated validator](https://docs.pydantic.dev/latest/concepts/validators/#annotated-validators):
```py3
from pydantic.functional_validators import AfterValidator

def do_something(v: str) -> str:
    if "something" in v:
        raise ValueError()
    return v

NonEmptyString = Annotated[
    constr(strip_whitespace=True, min_length=1, pattern='[^\0]+'),
    AfterValidator(do_something)
]
```
